File: public/js/indicators.js
/**
 * 기술적 지표 계산 모듈 (순수 함수).
 * 브라우저에서는 window.TA, Node 테스트에서는 module.exports 로 노출된다.
 * 모든 함수는 입력과 동일한 길이의 배열을 돌려주며, 계산 불가 구간은 null 이다.
 */
(function (root, factory) {
  const api = factory();
  if (typeof module === 'object' && module.exports) module.exports = api;
  else root.TA = api;
})(typeof self !== 'undefined' ? self : this, function () {
  'use strict';

  const num = (x) => (typeof x === 'number' && isFinite(x) ? x : null);

  /** 단순이동평균 (Simple Moving Average) */
  function sma(values, period) {
    const out = new Array(values.length).fill(null);
    if (period <= 0) return out;
    let sum = 0;
    let count = 0;
    for (let i = 0; i < values.length; i++) {
      const v = num(values[i]);
      if (v === null) {
        sum = 0;
        count = 0;
        continue;
      }
      sum += v;
      count++;
      if (count > period) {
        sum -= num(values[i - period]) || 0;
        count = period;
      }
      if (count === period) out[i] = sum / period;
    }
    return out;
  }

  /** 지수이동평균 (Exponential Moving Average) */
  function ema(values, period) {
    const out = new Array(values.length).fill(null);
    if (period <= 0) return out;
    const k = 2 / (period + 1);
    let prev = null;
    let seed = 0;
    let seen = 0;
    for (let i = 0; i < values.length; i++) {
      const v = num(values[i]);
      if (v === null) continue;
      if (prev === null) {
        seed += v;
        seen++;
        if (seen === period) {
          prev = seed / period;
          out[i] = prev;
        }
        continue;
      }
      prev = v * k + prev * (1 - k);
      out[i] = prev;
    }
    return out;
  }

  /** Wilder 평활 (RSI/ATR 등에서 사용) */
  function wilder(values, period) {
    const out = new Array(values.length).fill(null);
    let sum = 0;
    let seen = 0;
    let prev = null;
    for (let i = 0; i < values.length; i++) {
      const v = num(values[i]);
      if (v === null) continue;
      if (prev === null) {
        sum += v;
        seen++;
        if (seen === period) {
          prev = sum / period;
          out[i] = prev;
        }
        continue;
      }
      prev = (prev * (period - 1) + v) / period;
      out[i] = prev;
    }
    return out;
  }

  function stddev(values, period, means) {
    const out = new Array(values.length).fill(null);
    for (let i = period - 1; i < values.length; i++) {
      const mean = means[i];
      if (mean === null) continue;
      let acc = 0;
      let ok = true;
      for (let j = i - period + 1; j <= i; j++) {
        const v = num(values[j]);
        if (v === null) {
          ok = false;
          break;
        }
        acc += (v - mean) ** 2;
      }
      if (ok) out[i] = Math.sqrt(acc / period);
    }
    return out;
  }

  /**
   * 볼린저 밴드.
   * @returns {{mid:Array, upper:Array, lower:Array, width:Array, percentB:Array}}
   *  width  = (상단-하단)/중심선  → 스퀴즈(변동성 수축) 판정에 사용
   *  percentB = (종가-하단)/(상단-하단)
   */
  function bollinger(values, period = 20, mult = 2) {
    const mid = sma(values, period);
    const sd = stddev(values, period, mid);
    const upper = new Array(values.length).fill(null);
    const lower = new Array(values.length).fill(null);
    const width = new Array(values.length).fill(null);
    const percentB = new Array(values.length).fill(null);
    for (let i = 0; i < values.length; i++) {
      if (mid[i] === null || sd[i] === null) continue;
      upper[i] = mid[i] + mult * sd[i];
      lower[i] = mid[i] - mult * sd[i];
      width[i] = mid[i] ? (upper[i] - lower[i]) / mid[i] : null;
      const span = upper[i] - lower[i];
      percentB[i] = span ? (values[i] - lower[i]) / span : 0.5;
    }
    return { mid, upper, lower, width, percentB };
  }

  /** RSI (Wilder) */
  function rsi(values, period = 14) {
    const gains = new Array(values.length).fill(null);
    const losses = new Array(values.length).fill(null);
    for (let i = 1; i < values.length; i++) {
      const d = values[i] - values[i - 1];
      gains[i] = Math.max(0, d);
      losses[i] = Math.max(0, -d);
    }
    const ag = wilder(gains.slice(1), period);
    const al = wilder(losses.slice(1), period);
    const out = new Array(values.length).fill(null);
    for (let i = 0; i < ag.length; i++) {
      if (ag[i] === null || al[i] === null) continue;
      out[i + 1] = al[i] === 0 ? 100 : 100 - 100 / (1 + ag[i] / al[i]);
    }
    return out;
  }

  /** MACD (기본 12, 26, 9) */
  function macd(values, fast = 12, slow = 26, signal = 9) {
    const ef = ema(values, fast);
    const es = ema(values, slow);
    const line = values.map((_, i) => (ef[i] === null || es[i] === null ? null : ef[i] - es[i]));
    const compact = line.filter((v) => v !== null);
    const sigCompact = ema(compact, signal);
    const sig = new Array(values.length).fill(null);
    let k = 0;
    for (let i = 0; i < line.length; i++) {
      if (line[i] === null) continue;
      sig[i] = sigCompact[k++];
    }
    const hist = line.map((v, i) => (v === null || sig[i] === null ? null : v - sig[i]));
    return { line, signal: sig, hist };
  }

  /** 스토캐스틱 (%K, %D) */
  function stochastic(high, low, close, kPeriod = 14, kSmooth = 3, dPeriod = 3) {
    const raw = new Array(close.length).fill(null);
    for (let i = kPeriod - 1; i < close.length; i++) {
      let hh = -Infinity;
      let ll = Infinity;
      for (let j = i - kPeriod + 1; j <= i; j++) {
        hh = Math.max(hh, high[j]);
        ll = Math.min(ll, low[j]);
      }
      raw[i] = hh === ll ? 50 : ((close[i] - ll) / (hh - ll)) * 100;
    }
    const k = smoothNullable(raw, kSmooth);
    const d = smoothNullable(k, dPeriod);
    return { k, d };
  }

  function smoothNullable(arr, period) {
    const compact = [];
    const idx = [];
    for (let i = 0; i < arr.length; i++) {
      if (arr[i] !== null) {
        compact.push(arr[i]);
        idx.push(i);
      }
    }
    const sm = sma(compact, period);
    const out = new Array(arr.length).fill(null);
    for (let i = 0; i < idx.length; i++) out[idx[i]] = sm[i];
    return out;
  }

  /** True Range / ATR – 손절폭 산정에 사용 */
  function atr(high, low, close, period = 14) {
    const tr = new Array(close.length).fill(null);
    for (let i = 0; i < close.length; i++) {
      if (i === 0) {
        tr[i] = high[i] - low[i];
        continue;
      }
      tr[i] = Math.max(
        high[i] - low[i],
        Math.abs(high[i] - close[i - 1]),
        Math.abs(low[i] - close[i - 1])
      );
    }
    return wilder(tr, period);
  }

  /** 당일(세션) 기준 VWAP – 단타의 기준선 */
  function vwap(candles) {
    const out = new Array(candles.length).fill(null);
    let pv = 0;
    let vol = 0;
    let day = null;
    for (let i = 0; i < candles.length; i++) {
      const c = candles[i];
      const d = new Date(c.t).toISOString().slice(0, 10);
      if (d !== day) {
        day = d;
        pv = 0;
        vol = 0;
      }
      const typical = (c.h + c.l + c.c) / 3;
      const v = c.v || 0;
      pv += typical * v;
      vol += v;
      out[i] = vol > 0 ? pv / vol : typical;
    }
    return out;
  }

  /** OBV (누적 거래량) */
  function obv(candles) {
    const out = new Array(candles.length).fill(null);
    let acc = 0;
    for (let i = 0; i < candles.length; i++) {
      if (i > 0) {
        if (candles[i].c > candles[i - 1].c) acc += candles[i].v || 0;
        else if (candles[i].c < candles[i - 1].c) acc -= candles[i].v || 0;
      }
      out[i] = acc;
    }
    return out;
  }

  /**
   * 피봇(스윙) 고점/저점 탐지.
   * left/right 개의 봉보다 높은(낮은) 봉만 스윙으로 인정한다.
   */
  function pivots(candles, left = 3, right = 3) {
    const highs = [];
    const lows = [];
    for (let i = left; i < candles.length - right; i++) {
      let isHigh = true;
      let isLow = true;
      for (let j = i - left; j <= i + right; j++) {
        if (j === i) continue;
        if (candles[j].h >= candles[i].h) isHigh = false;
        if (candles[j].l <= candles[i].l) isLow = false;
        if (!isHigh && !isLow) break;
      }
      if (isHigh) highs.push({ i, price: candles[i].h, t: candles[i].t });
      if (isLow) lows.push({ i, price: candles[i].l, t: candles[i].t });
    }
    return { highs, lows };
  }

  /** 두 점을 잇는 직선 (y = slope * x + intercept, x는 봉 인덱스) */
  function lineThrough(a, b) {
    if (!a || !b || a.i === b.i) return null;
    const slope = (b.price - a.price) / (b.i - a.i);
    return { slope, intercept: a.price - slope * a.i, from: a, to: b };
  }

  /**
   * 자동 추세선: 최근 스윙 고점 2개(저항), 스윙 저점 2개(지지)를 연결.
   * 이후 봉에서 선을 뚫는 경우(돌파)를 신호 엔진이 사용한다.
   */
  function autoTrendlines(candles, left = 3, right = 3) {
    const { highs, lows } = pivots(candles, left, right);
    const pickHighs = highs.slice(-4);
    const pickLows = lows.slice(-4);
    const resistance = lineThrough(pickHighs[pickHighs.length - 2], pickHighs[pickHighs.length - 1]);
    const support = lineThrough(pickLows[pickLows.length - 2], pickLows[pickLows.length - 1]);
    return { resistance, support, highs, lows };
  }

  /** 최소제곱 회귀 추세선 (최근 n봉) */
  function regressionLine(values, lookback) {
    const n = Math.min(lookback, values.length);
    const start = values.length - n;
    let sx = 0;
    let sy = 0;
    let sxy = 0;
    let sxx = 0;
    let cnt = 0;
    for (let i = start; i < values.length; i++) {
      const y = num(values[i]);
      if (y === null) continue;
      const x = i;
      sx += x;
      sy += y;
      sxy += x * y;
      sxx += x * x;
      cnt++;
    }
    if (cnt < 2) return null;
    const denom = cnt * sxx - sx * sx;
    if (!denom) return null;
    const slope = (cnt * sxy - sx * sy) / denom;
    const intercept = (sy - slope * sx) / cnt;
    return { slope, intercept, from: { i: start }, to: { i: values.length - 1 } };
  }

  /** 지지·저항 레벨: 스윙 가격대를 근접도 기준으로 군집화하고 터치 횟수로 정렬 */
  function supportResistance(candles, tolerancePct = 0.35, maxLevels = 5, nearPct = 8) {
    const { highs, lows } = pivots(candles, 3, 3);
    const points = [...highs.map((p) => ({ ...p, kind: 'R' })), ...lows.map((p) => ({ ...p, kind: 'S' }))];
    const clusters = [];
    points.sort((a, b) => a.price - b.price);
    for (const p of points) {
      const last = clusters[clusters.length - 1];
      if (last && Math.abs(p.price - last.price) / last.price * 100 <= tolerancePct) {
        last.price = (last.price * last.touches + p.price) / (last.touches + 1);
        last.touches++;
        last.lastIndex = Math.max(last.lastIndex, p.i);
      } else {
        clusters.push({ price: p.price, touches: 1, lastIndex: p.i });
      }
    }
    // 현재가에서 너무 먼 레벨은 단타에 쓸모가 없으므로 제외
    const price = candles[candles.length - 1].c;
    const near = clusters
      .filter((c) => Math.abs(c.price - price) / price * 100 <= nearPct)
      .sort((a, b) => b.touches - a.touches || b.lastIndex - a.lastIndex);
    // 두 번 이상 눌린 자리를 우선하고, 부족하면 한 번짜리로 채운다
    const strong = near.filter((c) => c.touches >= 2);
    const picked = (strong.length >= 2 ? strong : near).slice(0, maxLevels);
    return picked.sort((a, b) => a.price - b.price);
  }

  /** 피보나치 되돌림 (최근 구간 고·저 기준) */
  function fibonacci(candles, lookback = 120) {
    const slice = candles.slice(-lookback);
    if (!slice.length) return null;
    let hi = -Infinity;
    let lo = Infinity;
    let hiIdx = 0;
    let loIdx = 0;
    slice.forEach((c, i) => {
      if (c.h > hi) {
        hi = c.h;
        hiIdx = i;
      }
      if (c.l < lo) {
        lo = c.l;
        loIdx = i;
      }
    });
    const up = loIdx < hiIdx; // 저점이 먼저면 상승 파동
    const diff = hi - lo;
    const levels = [0, 0.236, 0.382, 0.5, 0.618, 0.786, 1].map((r) => ({
      ratio: r,
      price: up ? hi - diff * r : lo + diff * r,
    }));
    return { high: hi, low: lo, up, levels };
  }

  /** 백분위 순위(0~100): 현재 값이 과거 구간에서 어느 위치인지 */
  function percentileRank(values, lookback) {
    const arr = values.slice(-lookback).filter((v) => v !== null && isFinite(v));
    if (arr.length < 5) return null;
    const cur = arr[arr.length - 1];
    const below = arr.filter((v) => v <= cur).length;
    return (below / arr.length) * 100;
  }

  /** 실현 변동성(연율화 %) – 참고 정보용 */
  function volatility(closes, period = 20, barsPerYear = 98280) {
    if (closes.length < period + 1) return null;
    const rets = [];
    for (let i = closes.length - period; i < closes.length; i++) {
      if (i <= 0) continue;
      rets.push(Math.log(closes[i] / closes[i - 1]));
    }
    const mean = rets.reduce((a, b) => a + b, 0) / rets.length;
    const varr = rets.reduce((a, b) => a + (b - mean) ** 2, 0) / (rets.length - 1 || 1);
    return Math.sqrt(varr) * Math.sqrt(barsPerYear) * 100;
  }

  return {
    sma, ema, wilder, bollinger, rsi, macd, stochastic, atr, vwap, obv,
    pivots, lineThrough, autoTrendlines, regressionLine, supportResistance,
    fibonacci, percentileRank, volatility,
  };
});

File: scalper/indicators.py
"""단타용 기술 지표 — 외부 의존성 없이 순수 파이썬으로 계산합니다.

모든 함수는 "과거 → 현재" 순서의 리스트를 받고, 계산 불가한 앞부분은 None 으로
채운 같은 길이의 리스트를 돌려줍니다. 인덱스가 캔들과 1:1 로 맞아야
차트/백테스트에서 미래참조(look-ahead) 사고가 나지 않습니다.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import Iterable, Sequence


@dataclass
class Candle:
    """5분봉 하나. ts 는 UTC epoch 초."""

    ts: int
    open: float
    high: float
    low: float
    close: float
    volume: float

    @property
    def is_bull(self) -> bool:
        return self.close >= self.open

    @property
    def body(self) -> float:
        return abs(self.close - self.open)

    @property
    def typical(self) -> float:
        return (self.high + self.low + self.close) / 3.0

    def as_dict(self) -> dict:
        return {
            "ts": self.ts,
            "o": round(self.open, 4),
            "h": round(self.high, 4),
            "l": round(self.low, 4),
            "c": round(self.close, 4),
            "v": round(self.volume, 2),
        }

    @classmethod
    def from_dict(cls, d: dict) -> "Candle":
        return cls(
            ts=int(d.get("ts") or d.get("t") or 0),
            open=float(d.get("o", d.get("open", 0.0))),
            high=float(d.get("h", d.get("high", 0.0))),
            low=float(d.get("l", d.get("low", 0.0))),
            close=float(d.get("c", d.get("close", 0.0))),
            volume=float(d.get("v", d.get("volume", 0.0))),
        )


Series = list[float | None]


def _f(values: Iterable[float | None]) -> list[float | None]:
    return [None if v is None else float(v) for v in values]


def sma(values: Sequence[float], period: int) -> Series:
    """단순이동평균."""
    out: Series = [None] * len(values)
    if period <= 0:
        return out
    acc = 0.0
    for i, v in enumerate(values):
        acc += v
        if i >= period:
            acc -= values[i - period]
        if i >= period - 1:
            out[i] = acc / period
    return out


def ema(values: Sequence[float], period: int) -> Series:
    """지수이동평균. 첫 값은 SMA 로 시드해 초기 왜곡을 줄입니다."""
    out: Series = [None] * len(values)
    if period <= 0 or len(values) < period:
        return out
    k = 2.0 / (period + 1.0)
    seed = sum(values[:period]) / period
    out[period - 1] = seed
    prev = seed
    for i in range(period, len(values)):
        prev = values[i] * k + prev * (1 - k)
        out[i] = prev
    return out


def rsi(values: Sequence[float], period: int = 14) -> Series:
    """와일더 방식 RSI."""
    out: Series = [None] * len(values)
    if len(values) <= period:
        return out
    gains = 0.0
    losses = 0.0
    for i in range(1, period + 1):
        diff = values[i] - values[i - 1]
        gains += max(diff, 0.0)
        losses += max(-diff, 0.0)
    avg_gain = gains / period
    avg_loss = losses / period
    out[period] = 100.0 if avg_loss == 0 else 100 - 100 / (1 + avg_gain / avg_loss)
    for i in range(period + 1, len(values)):
        diff = values[i] - values[i - 1]
        avg_gain = (avg_gain * (period - 1) + max(diff, 0.0)) / period
        avg_loss = (avg_loss * (period - 1) + max(-diff, 0.0)) / period
        out[i] = 100.0 if avg_loss == 0 else 100 - 100 / (1 + avg_gain / avg_loss)
    return out


def macd(
    values: Sequence[float], fast: int = 12, slow: int = 26, signal: int = 9
) -> tuple[Series, Series, Series]:
    """MACD 라인 / 시그널 / 히스토그램."""
    fast_line = ema(values, fast)
    slow_line = ema(values, slow)
    line: Series = [
        None if (f is None or s is None) else f - s
        for f, s in zip(fast_line, slow_line)
    ]
    dense = [v for v in line if v is not None]
    sig_dense = ema(dense, signal)
    sig: Series = [None] * len(values)
    offset = len(values) - len(dense)
    for i, v in enumerate(sig_dense):
        sig[offset + i] = v
    hist: Series = [
        None if (l is None or s is None) else l - s for l, s in zip(line, sig)
    ]
    return line, sig, hist


def bollinger(
    values: Sequence[float], period: int = 20, mult: float = 2.0
) -> tuple[Series, Series, Series, Series]:
    """볼린저밴드 (상단, 중심, 하단, 밴드폭%)."""
    mid = sma(values, period)
    upper: Series = [None] * len(values)
    lower: Series = [None] * len(values)
    width: Series = [None] * len(values)
    for i in range(len(values)):
        m = mid[i]
        if m is None:
            continue
        window = values[i - period + 1 : i + 1]
        var = sum((v - m) ** 2 for v in window) / period
        sd = math.sqrt(var)
        upper[i] = m + mult * sd
        lower[i] = m - mult * sd
        width[i] = (upper[i] - lower[i]) / m * 100 if m else None
    return upper, mid, lower, width


def vwap(candles: Sequence[Candle], session_reset: bool = True) -> Series:
    """거래량가중평균가. 단타의 기준선이라 세션(날짜)이 바뀌면 리셋합니다."""
    out: Series = [None] * len(candles)
    pv = 0.0
    vol = 0.0
    day = None
    for i, c in enumerate(candles):
        cur_day = c.ts // 86400
        if session_reset and day is not None and cur_day != day:
            pv = 0.0
            vol = 0.0
        day = cur_day
        pv += c.typical * c.volume
        vol += c.volume
        out[i] = pv / vol if vol > 0 else c.close
    return out


def atr(candles: Sequence[Candle], period: int = 14) -> Series:
    """평균실체범위 — 손절폭을 종목 변동성에 맞추는 데 씁니다."""
    out: Series = [None] * len(candles)
    if len(candles) < 2:
        return out
    trs: list[float] = [candles[0].high - candles[0].low]
    for i in range(1, len(candles)):
        prev_close = candles[i - 1].close
        c = candles[i]
        trs.append(
            max(c.high - c.low, abs(c.high - prev_close), abs(c.low - prev_close))
        )
    smoothed = sma(trs, period)
    for i, v in enumerate(smoothed):
        out[i] = v
    return out


def volume_ratio(candles: Sequence[Candle], period: int = 20) -> Series:
    """직전 period 평균 대비 현재 거래량 배수."""
    vols = [c.volume for c in candles]
    avg = sma(vols, period)
    out: Series = [None] * len(candles)
    for i, a in enumerate(avg):
        if a and a > 0:
            out[i] = vols[i] / a
    return out


def slope(values: Series, lookback: int = 3) -> float | None:
    """최근 lookback 구간의 기울기(단위: 값/봉). 추세 꺾임 감지용."""
    pts = [v for v in values[-(lookback + 1) :] if v is not None]
    if len(pts) < 2:
        return None
    return (pts[-1] - pts[0]) / (len(pts) - 1)


def percentile_rank(values: Series, value: float | None, window: int = 60) -> float | None:
    """최근 window 안에서 value 가 몇 % 지점인지. 볼린저 스퀴즈 판정에 사용."""
    if value is None:
        return None
    pool = [v for v in values[-window:] if v is not None]
    if len(pool) < 5:
        return None
    below = sum(1 for v in pool if v <= value)
    return below / len(pool) * 100.0


@dataclass
class Snapshot:
    """한 시점의 모든 지표를 한 번에 담아 신호 로직으로 넘깁니다."""

    price: float
    ma5: float | None = None
    ma20: float | None = None
    ma60: float | None = None
    ma120: float | None = None
    prev_ma5: float | None = None
    prev_ma20: float | None = None
    rsi: float | None = None
    prev_rsi: float | None = None
    macd: float | None = None
    macd_signal: float | None = None
    macd_hist: float | None = None
    prev_macd_hist: float | None = None
    bb_upper: float | None = None
    bb_mid: float | None = None
    bb_lower: float | None = None
    bb_width: float | None = None
    bb_width_pct: float | None = None
    vwap: float | None = None
    atr: float | None = None
    vol_ratio: float | None = None
    candles: list[Candle] = field(default_factory=list)

    def as_dict(self) -> dict:
        out = {}
        for k, v in self.__dict__.items():
            if k == "candles":
                continue
            out[k] = None if v is None else round(float(v), 4)
        return out


def compute(candles: Sequence[Candle]) -> Snapshot:
    """캔들 시퀀스 → 최신 시점 지표 스냅샷."""
    closes = [c.close for c in candles]
    if not closes:
        return Snapshot(price=0.0)
    ma5, ma20 = sma(closes, 5), sma(closes, 20)
    ma60, ma120 = sma(closes, 60), sma(closes, 120)
    rsi14 = rsi(closes, 14)
    line, sig, hist = macd(closes)
    up, mid, low, width = bollinger(closes, 20, 2.0)
    vw = vwap(candles)
    at = atr(candles, 14)
    vr = volume_ratio(candles, 20)
    i = len(candles) - 1
    j = i - 1 if i > 0 else i
    return Snapshot(
        price=closes[i],
        ma5=ma5[i],
        ma20=ma20[i],
        ma60=ma60[i],
        ma120=ma120[i],
        prev_ma5=ma5[j],
        prev_ma20=ma20[j],
        rsi=rsi14[i],
        prev_rsi=rsi14[j],
        macd=line[i],
        macd_signal=sig[i],
        macd_hist=hist[i],
        prev_macd_hist=hist[j],
        bb_upper=up[i],
        bb_mid=mid[i],
        bb_lower=low[i],
        bb_width=width[i],
        bb_width_pct=percentile_rank(width, width[i], 60),
        vwap=vw[i],
        atr=at[i],
        vol_ratio=vr[i],
        candles=list(candles),
    )
